crates/ark/src/reticulate.rs

```
@@ -0,0 +1,145 @@
+                log::error!("Connection Pane: Error while handling messages: {err:?}");
+            });
+        });
+
+        Ok(comm_id)
+    }
+
+    fn handle_messages(&self) -> Result<(), anyhow::Error> {
+        loop {
+            let msg = unwrap!(self.comm.incoming_rx.recv(), Err(err) => {
+                log::error!("Reticulate: Error while receiving message from frontend: {err:?}");
+                break;
+            });
+
+            if let CommMsg::Close = msg {
+                self.comm.outgoing_tx.send(CommMsg::Close).unwrap();
+                break;
+            }
+
+            // Forward data msgs to the frontend
+            if let CommMsg::Data(_) = msg {
+                self.comm.outgoing_tx.send(msg)?;
+                continue;
+            }
+        }
+
+        // before finalizing the thread we make sure to send a close message to the front end
+        if let Err(err) = self.comm.outgoing_tx.send(CommMsg::Close) {
+            log::error!("Reticulate: Error while sending comm_close to front end: {err:?}");
+        }
+
+        let mut comm_id_guard = unwrap!(
+            RETICULATE_COMM_ID.try_lock(),
+            Err(e) => {
+                return Err(anyhow!("Could not access comm_id. Error {}", e));
+            }
+        );
+        log::info!(
+            "Reticulate Thread closing {}",
+            (*comm_id_guard).clone().unwrap()
+        );
+
+        *comm_id_guard = None;
+
+        Ok(())
+    }
+}
+
+// Creates a client instance reticulate can use to communicate with the front-end.
+// We should aim at having at most **1** client per R session.
+// Further actions that reticulate can ask the front-end can be requested through
+// the comm_id that is returned by this function.
+#[harp::register]
+pub unsafe extern "C" fn ps_reticulate_open(input: SEXP) -> Result<SEXP, anyhow::Error> {
+    let main = RMain::get();
+
+    if !RMain::initialized() {
+        return Ok(R_NilValue);
+    }
```

This should probably surface an error.

But once https://github.com/posit-dev/ark/pull/542 is merged, you can use `RMain::wait_initialized()`. If you need R to be ready too, you can also use `RMain::wait_r_initialized()`.

That said this method is called from R so actually this probably should be removed altogether.

---

This was required for testing the connections comm, because loading a package could call comm_open, but it's not required here and I accidentally copied it over.